src/hotspot/share/gc/g1/g1EvacFailureRegions.cpp

```diff
@@ -24,47 +24,132 @@
+  assert(_chunks_in_regions == nullptr, "not cleaned up");
 }
 
 void G1EvacFailureRegions::pre_collection(uint max_regions) {
   Atomic::store(&_evac_failure_regions_cur_length, 0u);
-  _max_regions = max_regions;
-  _regions_failed_evacuation.resize(_max_regions);
-  _evac_failure_regions = NEW_C_HEAP_ARRAY(uint, _max_regions, mtGC);
+  _regions_failed_evacuation.resize(max_regions);
+  _evac_failure_regions = NEW_C_HEAP_ARRAY(uint, max_regions, mtGC);
+  _chunks_in_regions = new (NEW_C_HEAP_OBJ(G1ScanChunksInHeapRegions, mtGC)) G1ScanChunksInHeapRegions();
 }
 
 void G1EvacFailureRegions::post_collection() {
   _regions_failed_evacuation.resize(0);
+
+  FREE_C_HEAP_OBJ(_chunks_in_regions);
+  _chunks_in_regions = nullptr;
+
   FREE_C_HEAP_ARRAY(uint, _evac_failure_regions);
   _evac_failure_regions = nullptr;
-  _max_regions = 0; // To have any record() attempt fail in the future.
+}
+
+bool G1EvacFailureRegions::contains(uint region_idx) const {
+  return _regions_failed_evacuation.par_at(region_idx, memory_order_relaxed);
 }
 
 void G1EvacFailureRegions::par_iterate(HeapRegionClosure* closure,
-                                       HeapRegionClaimer* _hrclaimer,
+                                       HeapRegionClaimer* hrclaimer,
                                        uint worker_id) const {
   G1CollectedHeap::heap()->par_iterate_regions_array(closure,
-                                                     _hrclaimer,
+                                                     hrclaimer,
                                                      _evac_failure_regions,
                                                      Atomic::load(&_evac_failure_regions_cur_length),
                                                      worker_id);
 }
 
-bool G1EvacFailureRegions::contains(uint region_idx) const {
-  assert(region_idx < _max_regions, "must be");
-  return _regions_failed_evacuation.par_at(region_idx, memory_order_relaxed);
+void G1EvacFailureRegions::initialize_chunks(uint num_workers, const char* task_name) {
+  _chunks_in_regions->initialize(_evac_failure_regions,
+                                 Atomic::load(&_evac_failure_regions_cur_length),
+                                 num_workers, task_name);
+}
+
+void G1EvacFailureRegions::par_iterate_chunks_in_regions(G1HeapRegionChunkClosure* chunk_closure,
+                                                         uint worker_id) const {
+  _chunks_in_regions->par_iterate_chunks_in_regions(chunk_closure, worker_id);
+}
+
+class PrepareEvacFailureRegionTask : public WorkerTask {
+  G1EvacFailureRegions* _evac_failure_regions;
+  uint _num_workers;
+  HeapRegionClaimer _claimer;
+
+  class PrepareEvacFailureRegionClosure : public HeapRegionClosure {
+    const G1EvacFailureRegions* _evac_failure_regions;
+    uint _worker_id;
+
+    void prepare_region(uint region_idx, uint worker_id) {
+      G1CollectedHeap* _heap = G1CollectedHeap::heap();
```

Comment:
Locals should not have an underscore. I personally also prefer `g1h` for `G1CollectedHeap::heap()`, but `heap` is fine.